### ISSUE TITLE
Improve the `The pipe could not be found` error message

### ISSUE THREAD

**Issue by AndrewKushnir** [angular/angular#43855]
### Which @angular/* package(s) are relevant/releated to the feature request?

core

### Description

Currently Angular throws the following error message at runtime when a pipe can not be found:

```
The pipe `AsyncPipe` could not be found!
```

However there is no information on which component has this problem, which makes debugging harder. We should consider adding component class name to the error message.

**Comment by rohan-pednekar** [comment]
Sure, @AndrewKushnir. Sorry for the late response due to a tight schedule. I'll try my best to fix this issue as I get free time. If I encounter any problem, I'll let you know here or on the channel. Thank you.

**Comment by AndrewKushnir** [comment]
@rohan-pednekar thanks for the reply. If you don't have a chance to help - don't worry, I can take care of creating a PR and you can help with other fixes/improvements in the future. Thank you.

**Comment by rohan-pednekar** [comment]
@AndrewKushnir is there a way to run only one spec file ? I tried `fdescribe`, but it seems its not working...

**Comment by rohan-pednekar** [comment]
Hi @AndrewKushnir This is the test suite that I have added in the `pipe_spec.ts` and its passing (Used this command to test only render3 test suite)

 ```
yarn bazel test packages/core/test/render3:render3 --test_output=streamed
```
Is this correct ? 

(P.S: As I am contributing for the first time, please correct me if I am doing something wrong. Thank you.)

```js
  describe('should throw error for missing pipe', () => {
    it('should throw error for missing pipe in TestComponent', () => {
      @Component({
        template: `
          {{ value | testMissingPipe }}`
      })
      
      class TestComponent {
        value: string = 'test';
      }
  
      TestBed.configureTestingModule({declarations: [TestComponent]});
  
      expect(() => {
        const fixture = TestBed.createComponent(TestComponent);
        fixture.detectChanges();
      }).toThrowError(/The pipe 'testMissingPipe' could not be found  in the 'TestComponent' component!/);
    });

    it('should throw error for missing pipe in TestComponent', () => {
      @Component({
        template: `
          <ng-container *ngIf="true">
            {{ value | testMissingPipe }}
          </ng-container>`
      })

      class TestComponent {
        value: string = 'test';
      }
  
      TestBed.configureTestingModule({declarations: [TestComponent]});
  
      expect(() => {
        const fixture = TestBed.createComponent(TestComponent);
        fixture.detectChanges();
      }).toThrowError(/The pipe 'testMissingPipe' could not be found  in the 'TestComponent' component!/);
    });


    it('should throw error for missing pipe in TestComponent', () => {
      @Component({
        template: `
          <test-component>
            {{ value | testMissingPipe }}
          </test-component>`
      })

      class TestComponent {
        value: string = 'test';
      }
  
      TestBed.configureTestingModule({declarations: [TestComponent]});
  
      expect(() => {
        const fixture = TestBed.createComponent(TestComponent);
        fixture.detectChanges();
      }).toThrowError(/The pipe 'testMissingPipe' could not be found  in the 'TestComponent' component!/);
    });

    it('should throw error for missing pipe in TestComponent', () => {
      @Component({
        template: `
          <test-component>
            <ng-container *ngIf="true">
              {{ value | testMissingPipe }}
            </ng-container>
          </test-component>`
      })

      class TestComponent {
        value: string = 'test';
      }
  
      TestBed.configureTestingModule({declarations: [TestComponent]});
  
      expect(() => {
        const fixture = TestBed.createComponent(TestComponent);
        fixture.detectChanges();
      }).toThrowError(/The pipe 'testMissingPipe' could not be found  in the 'TestComponent' component!/);
    });

    it('should throw error for missing pipe in TestComponent', () => {
      @Component({
        template: `
          <div *ngIf="isVisible | testMissingPipe"></div>`
      })

      class TestComponent {
        isVisible: boolean = true;
      }
  
      TestBed.configureTestingModule({declarations: [TestComponent]});
  
      expect(() => {
        const fixture = TestBed.createComponent(TestComponent);
        fixture.detectChanges();
      }).toThrowError(/The pipe 'testMissingPipe' could not be found  in the 'TestComponent' component!/);
    });

  });
```

**Comment by rohan-pednekar** [comment]
One more thing, as this test isn't in `acceptance` shall I keep it or remove it ?

```js
  // This test isn't in `acceptance`, because we can't capture the same behavior that we want
  // when going through `TestBed`. Here we're testing the behavior of AOT-compiled code which
  // differs from the JIT code in `TestBed`, because it includes a `ɵɵgetInheritedFactory` call
  // when the pipe is using inheritance.
  it('should be able to use DI in a Pipe that extends an Injectable', () => {
    @Injectable({providedIn: 'root'})
    class SayHelloService {
      getHello() {
        return 'Hello there';
      }
      static ɵfac = () => new SayHelloService();
      static ɵprov = ɵɵdefineInjectable(
          {token: SayHelloService, factory: SayHelloService.ɵfac, providedIn: 'root'});
    }

    @Injectable()
    class ParentPipe {
      constructor(protected sayHelloService: SayHelloService) {}
      static ɵfac = (t?: any) => new(t || ParentPipe)(ɵɵinject(SayHelloService));
      static ɵprov = ɵɵdefineInjectable({token: ParentPipe, factory: ParentPipe.ɵfac});
    }

    @Pipe({name: 'sayHello', pure: true})
    class SayHelloPipe extends ParentPipe implements PipeTransform {
      transform() {
        return this.sayHelloService.getHello();
      }
      static override ɵfac = (t?: any) => ɵɵgetInheritedFactory(t || SayHelloPipe)(SayHelloPipe);
      static ɵpipe = ɵɵdefinePipe({name: 'sayHello', type: SayHelloPipe, pure: true});
    }

    const fixture = new TemplateFixture({
      create: () => {
        ɵɵtext(0);
        ɵɵpipe(1, 'sayHello');
      },
      update: () => {
        ɵɵtextInterpolate1('', ɵɵpipeBind1(1, 1, null), '');
      },
      decls: 2,
      vars: 3,
      pipes: [SayHelloPipe]
    });

    expect(fixture.html).toBe('Hello there');
  });
```

**Comment by rohan-pednekar** [comment]
Hi @AndrewKushnir I am trying to build and test all packages but, I am getting the following error...
<br>
![error](https://user-images.githubusercontent.com/80909812/140032173-78c4841c-df9c-4d49-8bb7-96051507cb69.PNG)

**Comment by AndrewKushnir** [comment]
@rohan-pednekar I think you should add tests to the `packages/core/test/acceptance/pipe_spec.ts` file only, there is no need to add them to the `render3` directory (we would migrate the remaining tests from there to the `acceptance` folder eventually). Once the tests are added, you can use the following command to run them:
```
yarn bazel test --config=ivy packages/core/test/acceptance
```
You could also focus a particular suite or a test using `fdescribe` or `fit`.

Please let me know if that worked.

Thank you.

**Comment by rohan-pednekar** [comment]
@AndrewKushnir  Yes above command is working...
<br>
![success-200](https://user-images.githubusercontent.com/80909812/140266422-9900e61f-edd4-492c-9c84-27c7054b12e1.png)

**Comment by rohan-pednekar** [comment]
> Hi @AndrewKushnir I am trying to build and test all packages but, I am getting the following error... ![error_ new](https://user-images.githubusercontent.com/80909812/140045492-b508ba35-2992-4f8b-9139-bf9d83ebc53d.PNG)

Tried below command to test all the packages but still its failing and I am getting the same error.
```
yarn bazel test --config=ivy packages/...
```

**Comment by AndrewKushnir** [comment]
@rohan-pednekar we can try to reproduce the problem that you reported, but since the `acceptance` tests are successful, this is a good signal and you can proceed with creating a PR (so you are not blocked). Once you create a PR, the CI will run all the necessary tests and if there are any issues, we'll investigate more.

**Comment by rohan-pednekar** [comment]
Thank you, @AndrewKushnir.  Sure, I'll create a PR. Also, I have tried formatting all the source code. But, I am seeing changes in many files that I haven't even touched. I used the following command.
```
yarn ng-dev format all
```
So am I supposed to format only those files in which I did changes or need to format all the source code?

**Comment by rohan-pednekar** [comment]
> Hi @AndrewKushnir I am trying to build and test all packages but, I am getting the following error... ![error_ new](https://user-images.githubusercontent.com/80909812/140045492-b508ba35-2992-4f8b-9139-bf9d83ebc53d.PNG)

I have tried many ways to resolve this issue, but I think it's a platform-specific `rollup_bundle` issue in Bazel (as I am using a Windows machine).

**Comment by AndrewKushnir** [comment]
> I have tried formatting all the source code. But, I am seeing changes in many files that I haven't even touched. I used the following command.

You could target changed files only using this command:

```
yarn ng-dev format changed
```

More info can be found at https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#formatting-your-source-code.

**Comment by rohan-pednekar** [comment]
@AndrewKushnir tests failed on CircleCI. Need to add template for `TestChildComponent`. Also, its throwing a `Template parse errors`. I'll update the code and will create a fixup commit. Thank you.

**Comment by rohan-pednekar** [comment]
Hi @AndrewKushnir What do I need to do in this case? As Type 'undefined' is not assignable to type 'number | PipeDef<any> | TNode | DirectiveDef<any> | ComponentDef<any> | TStylingRange | TStylingKey | ProviderToken<...> | TI18n | I18nUpdateOpCodes | TIcu'. for `tView.data[adjustedIndex]`.

<br>
<img width="789" alt="Screenshot 2021-11-09 at 8 36 48 PM" src="https://user-images.githubusercontent.com/80909812/140950467-c3560196-71f8-40f1-8f51-243afac25f89.png">

<br>
Also need to add check for `pipeDef`

```js
if(pipeDef && pipeDef.onDestroy) {

}
```

Is this correct ?

**Comment by AndrewKushnir** [comment]
@rohan-pednekar please see my comment to your PR: https://github.com/angular/angular/pull/44081#discussion_r743914851. Here is the change that should take care of the mentioned issue:
```
   if (tView.firstCreatePass) {
-    pipeDef = getPipeDef(pipeName, tView.pipeRegistry);
+    // The `getPipeDef` throws if a pipe with a given name is not found
+    // (so we use non-null assertion below).
+    pipeDef = getPipeDef(pipeName, tView.pipeRegistry)!;

```

**Comment by rohan-pednekar** [comment]
> @AndrewKushnir tests failed on **_legacy-unit-tests-saucelabs_ - CircleCI**. Need to add template for `TestChildComponent`. Also, its throwing a `Template parse errors`. I'll update the code and will create a fixup commit. Thank you.

Hi @AndrewKushnir Sorry for that lint error. But still test are failing due to Template Parse Error. Shall I add `schemas: [NO_ERRORS_SCHEMA]` in the `TestBed.configureTestingModule` ?

**Comment by AndrewKushnir** [comment]
Resolved by #44081.
